Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabSystemComponent.cpp

```diff
@@ -153,6 +153,24 @@ namespace AzToolsFramework
             }
         }
 
+        void PrefabSystemComponent::PropagateTemplateChangesDown(TemplateId templateId, InstanceOptionalReference instanceToExclude)
+        {
+            PropagateTemplateChanges(templateId, instanceToExclude);
+
+            auto templateIterator = m_templateIdMap.find(templateId);
+            if (templateIterator != m_templateIdMap.end())
+            {
+                for (LinkId linkId : templateIterator->second.GetLinks())
+                {
+                    auto linkIterator = m_linkIdMap.find(linkId);
+                    if (linkIterator != m_linkIdMap.end())
+                    {
+                        PropagateTemplateChangesDown(linkIterator->second.GetSourceTemplateId(), instanceToExclude);
+                    }
+                }
+            }
+        }
```

Comment:
PropagateTemplateChanges, via UpdatePrefabInstances, add the instances to the propagation queue.

In the case of a nested prefab hierarchy, before this change only the instance where a change happened would be refreshed - and the serializer would refresh it and recursively refresh all its children too. With this change, you will also add all children instances to the queue, re-refreshing them and turning the refresh from linear to quadratic.

If the current propagation code has issues maintaining links, we may have to have a better look at the Instance Serializer whenever nested prefabs are refreshed.

Reply:
I am not confident about the fix either, but I've just started learning the code, links, templates and all this stuff. Do you suggest looking elsewhere for a fix? From what I've managed to find out debugging is that when a level is loaded with a prefab that has nested prefab(s), the nested prefab instances themselves aren't getting linkId set. So when nested prefab is unparented, its linkid is still invalid causing a crash in PrefabPublicHandler::RemoveLink.

At the moment the function I've added is only getting called once upon level load so it propagates and updates everything, not just immediate children of level prefab but their children as well.

Perhaps linkId's should be determined lazily when we actually need them?

Reply:
I've talked to Ram and he proposed this change (see update)